prusti-specs/src/rewriter.rs

```diff
@@ -209,6 +223,159 @@ impl AstRewriter {
+            pledge_rhs(result, #rhs)
+        }
+    }
+}
+
+pub fn translate_pledge(reference: Option<syn::Expr>, lhs: syn::Expr, rhs: syn::Expr) -> syn::Expr {
+    if let Some(reference) = reference {
+        parse_quote_spanned! {
+            // TODO: figure out span
+            Span::call_site() =>
+            pledge(#reference, #lhs, #rhs)
+        }
+    } else {
+        parse_quote_spanned! {
+            // TODO: figure out span
+            Span::call_site() =>
+            pledge(result, #lhs, #rhs)
+        }
+    }
+}
+
+pub fn translate_implication(lhs: syn::Expr, rhs: syn::Expr) -> syn::Expr {
+    parse_quote_spanned! {
+        lhs.span().join(rhs.span()).unwrap() =>
+        implication(#lhs, #rhs)
+    }
+}
+
+pub fn translate_conjunction(mut conjuncts: Vec<syn::Expr>) -> syn::Expr {
+    debug_assert!(conjuncts.len() != 0, "empty conjuncts given");
+    if conjuncts.len() == 1 {
+        conjuncts.pop().unwrap()
+    } else {
+        let last = conjuncts.pop().unwrap();
+        let rest = translate_conjunction(conjuncts);
+        parse_quote_spanned! {
+            rest.span().join(last.span()).unwrap() =>
+            #rest && #last
+        }
+    }
+}
+
+fn args_to_tokens(mut args: Vec<(syn::Ident, syn::Type)>) -> TokenStream {
+    if args.len() == 0 {
+        TokenStream::new()
+    } else if args.len() == 1 {
+        let (ident, typ) = args.pop().unwrap();
+        quote_spanned! {
+            ident.span().join(typ.span()).unwrap() =>
+            #ident: #typ
+        }
+    } else {
+        let (ident, typ) = args.pop().unwrap();
+        let rest = args_to_tokens(args);
+        quote_spanned! {
+            rest.span().join(typ.span()).unwrap() =>
+            #rest, #ident: #typ
+        }
+    }
+}
+
+pub fn translate_spec_entailment(closure: syn::Expr, args: Vec<(syn::Ident, syn::Type)>, pres: Vec<syn::Expr>, posts: Vec<syn::Expr>) -> syn::Expr {
+    let arg_tokens = args_to_tokens(args);
+    let pre_conjuncts = translate_conjunction(pres);
+    let post_conjuncts = translate_conjunction(posts);
+    parse_quote_spanned! {
+        // TODO: get the right span
+        Span::call_site() =>
+        entailment(#closure, |#arg_tokens| {
+            {
+                #pre_conjuncts
+            } &&
+            {
+                #post_conjuncts
+            }
+        })
     }
 }
+
+fn tuple_to_tokens(mut exprs: Vec<syn::Expr>) -> TokenStream {
+    if exprs.len() == 0 {
+        quote_spanned! {
+            // TODO: get the right span
+            Span::call_site() =>
+            ,
+        }
+    } else if exprs.len() == 1 {
+        let expr = exprs.pop().unwrap();
+        quote_spanned! {
+            expr.span() =>
+            #expr,
+        }
+    } else {
+        let last = exprs.pop().unwrap();
+        let rest = tuple_to_tokens(exprs);
+        quote_spanned! {
+            rest.span().join(last.span()).unwrap() =>
+            #rest #last,
+        }
+    }
```

Comment:
`reduce`?

Reply:
Fixed in 6268917f30bbc875485f34bde92f5bd18cbde277